c_src/main.c

```diff
@@ -75,27 +75,19 @@ static void
 erlfdb_future_cb(FDBFuture* fdb_future, void* data)
 {
     ErlFDBFuture* future = (ErlFDBFuture*) data;
-    ErlNifEnv* caller;
     ERL_NIF_TERM msg;
+    bool cancelled;
 
-    // FoundationDB callbacks can fire from the thread
-    // that created them. Check if we were actually
-    // submitted to the network thread or not so that
-    // we pass the correct environment to enif_send
-    if(enif_thread_type() == ERL_NIF_THR_UNDEFINED) {
-        caller = NULL;
-    } else {
-        caller = future->pid_env;
-    }
 
     enif_mutex_lock(future->lock);
+    cancelled = future->cancelled;
+    enif_mutex_unlock(future->lock);
 
-    if(!future->cancelled) {
+    if(!cancelled) {
         msg = T2(future->msg_env, future->msg_ref, ATOM_ready);
-        enif_send(caller, &(future->pid), future->msg_env, msg);
+        enif_send(NULL, &(future->pid), future->msg_env, msg);
```

Comment:
we need to check the return value because if send failed, we will have some garbage in the `msg_env` that need to be cleaned otherwise memory leak.

Reply:
There is a resource destructor that makes cleanup unconditionally: https://github.com/savonarola/couchdb-erlfdb/blob/5636dc895a1431ca725831f4b4b5dcd8b1d2c38d/c_src/resources.c#L87